Graphics/GraphicsEngineVulkan/src/VulkanUtilities/VulkanLogicalDevice.cpp

```diff
@@ -31,14 +31,495 @@
 {
 
-std::shared_ptr<VulkanLogicalDevice> VulkanLogicalDevice::Create(VkPhysicalDevice             vkPhysicalDevice,
+// KHR ray tracing emulation.
+// Will be deprecated after the release of KHR extension.
+#if DILIGENT_USE_VOLK
+static_assert(sizeof(VkAccelerationStructureKHR) == sizeof(VkAccelerationStructureNV), "not compatible with NV extension");
+static_assert(sizeof(VkDeviceAddress) == 8, "not compatible with NV extension");
+
+static std::mutex                                    g_BufferDeviceAddressGuard;
+static std::unordered_map<VkDeviceAddress, VkBuffer> g_DeviceAddressToBuffer;
+static std::unordered_map<VkBuffer, VkDeviceAddress> g_BufferToDeviceAddress;
+static uint32_t                                      g_BufferDeviceAddressCounter = 0;
+static const VkDeviceAddress                         g_BufferMask                 = 0xFFFFFFFF00000000ull;
+
+static PFN_vkCreateBuffer              Origin_vkCreateBuffer              = nullptr;
+static PFN_vkDestroyBuffer             Origin_vkDestroyBuffer             = nullptr;
+static PFN_vkGetBufferDeviceAddressKHR Origin_vkGetBufferDeviceAddressKHR = nullptr;
+
+static VkResult VKAPI_CALL Wrap_vkCreateBuffer(VkDevice                     device,
+                                               const VkBufferCreateInfo*    pCreateInfo,
+                                               const VkAllocationCallbacks* pAllocator,
+                                               VkBuffer*                    pBuffer)
+{
+    const_cast<VkBufferCreateInfo*>(pCreateInfo)->usage &= ~VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT;
+    return Origin_vkCreateBuffer(device, pCreateInfo, pAllocator, pBuffer);
+}
+
+static void VKAPI_CALL Wrap_vkDestroyBuffer(VkDevice                     device,
+                                            VkBuffer                     buffer,
+                                            const VkAllocationCallbacks* pAllocator)
+{
+    Origin_vkDestroyBuffer(device, buffer, pAllocator);
+
+    std::unique_lock<std::mutex> lock{g_BufferDeviceAddressGuard};
+
+    auto iter = g_BufferToDeviceAddress.find(buffer);
+    if (iter != g_BufferToDeviceAddress.end())
+    {
+        g_DeviceAddressToBuffer.erase(iter->second);
+        g_BufferToDeviceAddress.erase(iter);
+    }
+}
+
+static VkDeviceAddress VKAPI_CALL Wrap_vkGetBufferDeviceAddressKHR(VkDevice                         device,
+                                                                   const VkBufferDeviceAddressInfo* pInfo)
+{
+    VERIFY_EXPR(pInfo->sType == VK_STRUCTURE_TYPE_BUFFER_DEVICE_ADDRESS_INFO_KHR);
+    VERIFY_EXPR(pInfo->pNext == nullptr);
+
+    std::unique_lock<std::mutex> lock{g_BufferDeviceAddressGuard};
+
+    // find in existing buffers
+    auto iter = g_BufferToDeviceAddress.find(pInfo->buffer);
+    if (iter != g_BufferToDeviceAddress.end())
+        return iter->second;
+
+    // create new device address
+    VkDeviceAddress Addr = VkDeviceAddress(++g_BufferDeviceAddressCounter) << 32;
```

Comment:
What does this do? If this is just a fake address, can you simply use buffer handle instead?

Reply:
VkDeviceAddress  is 64 bit value that stores GPU memory addres of VkBuffer + offset, I can not pack 64 bit handle and 64 bit offset into 64 bit.